asset/group_key_reveal_test.go

```diff
@@ -0,0 +1,144 @@
+package asset
+
+import (
+	"bytes"
+	"testing"
+
+	"github.com/btcsuite/btcd/btcec/v2"
+	"github.com/btcsuite/btcd/chaincfg/chainhash"
+	"github.com/btcsuite/btcd/txscript"
+	"github.com/lightninglabs/taproot-assets/fn"
+	"github.com/lightninglabs/taproot-assets/internal/test"
+	"github.com/stretchr/testify/require"
+)
+
+type testCaseGkrEncodeDecode struct {
+	testName string
+
+	internalKey       btcec.PublicKey
+	genesisAssetID    ID
+	customSubtreeRoot fn.Option[chainhash.Hash]
+}
+
+// GroupKeyReveal generates a GroupKeyReveal instance from the test case.
+func (tc testCaseGkrEncodeDecode) GroupKeyReveal() (GroupKeyReveal, error) {
+	gkr, err := NewGroupKeyRevealV1(
+		tc.internalKey, tc.genesisAssetID, tc.customSubtreeRoot,
+	).Unpack()
+
+	return &gkr, err
+}
+
+func TestGroupKeyTapscriptEncodeDecode(t *testing.T) {
```

Comment:
Here's a quick mafs version (may need some tweaking) of this test that uses `rapid`: 
```go
func TestGroupKeyTapscriptEncodeDecode(tt *testing.T) {
	ttt.Parallel()

	rapid.Check(tt, func(t *rapid.T) {
		// Generate random test inputs using rapid generators
		internalKey := *rapid.Custom(func(t *rapid.T) *btcec.PublicKey {
			priv, err := btcec.NewPrivateKey()
			if err != nil {
				t.Fatal(err)
			}
			return priv.PubKey()
		}).Draw(t, "internal_key")

		genesisAssetID := ID(rapid.SliceOfN([]byte{}, 32, 32).Draw(t, "genesis_id"))

		// Randomly decide whether to include a custom script
		hasCustomScript := rapid.Bool().Draw(t, "has_custom_script")

		var customSubtreeRoot fn.Option[chainhash.Hash]
		var customScriptLeaf *txscript.TapLeaf

		if hasCustomScript {
			// Generate random script between 1-100 bytes
			scriptSize := rapid.IntRange(1, 100).Draw(t, "script_size")
			customScript := rapid.SliceOfN([]byte{}, scriptSize, scriptSize).Draw(t, "custom_script")
			
			customScriptLeaf = txscript.NewBaseTapLeaf(customScript)
			customSubtreeRoot = fn.Some(customScriptLeaf.TapHash())
		} else {
			customSubtreeRoot = fn.None[chainhash.Hash]()
		}

		// Create and test GroupKeyReveal
		gkr, err := NewGroupKeyRevealV1(
			internalKey, 
			genesisAssetID,
			customSubtreeRoot,
		).Unpack()
		require.NoError(t, err)

		groupPubKey, err := gkr.GroupPubKey(genesisAssetID)
		require.NoError(t, err)

		// Test encoding/decoding
		var buffer bytes.Buffer
		var scratchBuffEncode [8]byte
		err = GroupKeyRevealEncoder(&buffer, &gkr, &scratchBuffEncode)
		require.NoError(t, err)

		var gkrDecoded GroupKeyReveal
		var scratchBuffDecode [8]byte
		err = GroupKeyRevealDecoder(
			&buffer,
			&gkrDecoded,
			&scratchBuffDecode,
			uint64(buffer.Len()),
		)
		require.NoError(t, err)

		// Prepare for comparison by removing non-encoded fields
		gkrV1, ok := gkr.(*GroupKeyRevealV1)
		require.True(t, ok)
		gkrV1.tapscript.customSubtreeInclusionProof = nil

		// Compare decoded with original
		require.Equal(t, gkrV1, gkrDecoded)

		// Verify decoded group public key
		groupPubKeyDecoded, err := gkrDecoded.GroupPubKey(genesisAssetID)
		require.NoError(t, err)
		require.Equal(t, groupPubKey, groupPubKeyDecoded)

		// Test custom subtree if present
		if customSubtreeRoot.IsSome() && customScriptLeaf != nil {
			gkrDecodedV1, ok := gkrDecoded.(*GroupKeyRevealV1)
			require.True(t, ok)

			ctrlBlock, err := gkrDecodedV1.ScriptSpendControlBlock(
				genesisAssetID,
			).Unpack()
			require.NoError(t, err)

			computedRoot := chainhash.Hash(
				ctrlBlock.RootHash(customScriptLeaf.Script),
			)
			require.Equal(t, gkrDecodedV1.tapscript.root, computedRoot)
		}
	})
}
```

Could be made more succinct with custom generators.

Reply:
Latest commit will include this test :+1: Thanks!